commands/previewPush.go

```diff
@@ -216,6 +235,59 @@ DomainLoop:
 		totalCorrections += len(corrections)
 		anyErrors = printOrRunCorrections(domain.Name, domain.RegistrarName, corrections, out, push, interactive, notifier) || anyErrors
 	}
+
+	if args.DeleteUnmanaged || args.DeleteUnmanagedDomains {
+		fmt.Printf("Checking Domain Removal:\n")
+		for registrarName, registrar := range createdRegistrars {
+			domainLister, ok := registrar.(providers.DomainLister)
+			if !ok {
+				out.Warnf("provider type of %s cannot list domains\n", registrarName)
+				continue
+			}
+			deployedDomains, err := domainLister.ListDomains()
+			if err != nil {
+				return fmt.Errorf("failed ListZones: %w\n", err)
+			}
+			for _, domain := range deployedDomains {
+				if !IsDomainManagedByRegistrar(cfg, domain, registrarName) {
+					fmt.Printf("Removing domain %s from provider %s\n", domain, registrarName)
+					totalCorrections += 1
+					if remover, ok := registrar.(providers.DomainRemover); ok && push {
+						if err := remover.EnsureDomainAbsent(domain); err != nil {
+							out.Errorf("Error deleting domain: %s\n", err)
+						}
+					}
+				}
+			}
+		}
+	}
+
+	if args.DeleteUnmanaged || args.DeleteUnmanagedZones {
+		fmt.Printf("Checking Zone Removal:\n")
+
+		for providerName, provider := range createdProviders {
+			zoneLister, ok := provider.(providers.ZoneLister)
+			if !ok {
+				out.Warnf("provider type of %s cannot list zones\n", providerName)
+				continue
+			}
+			deployedZones, err := zoneLister.ListZones()
+			if err != nil {
+				return fmt.Errorf("failed ListZones: %w\n", err)
+			}
+			for _, zone := range deployedZones {
+				if !IsZoneManagedByProvider(cfg, zone, providerName) {
+					fmt.Printf("Removing zone %s from provider %s\n", zone, providerName)
```

Comment:
Sadly CSCGLOBAL (and others?) are showing every domain as unmanaged.

I think the problem is that the code should compare deployedZone to the list of zones that are in dnsconfig.js (`cfg.Domains`).

Suggestion 1:
Turn this into a function. All the code included by `if DeleteUnmanaged || args.DeleteUnmanagedZones {` should be its own function.  Rather than including complex algorithms in the middle of an already long function, separate it out to make it easier to read and test.

Suggestion 2:
My other suggestion is to use "sets" instead of iteration.

Step 1: Extract the data you need into 3 lists:

```
configList := // zone names in dnsconfig.js. The `.Name` of each `cfg.Domains` item.
providerZoneList := // The zones that ListZones() output
providerDomainList := // The zones that ListDomains() output
```

Step 2: Now generate the list of zones that are unmanaged:

```
unmanaged_zones := providerZoneList - configList
unmanaged_domains := providerDomainList - configList
```

(In these formula `-` means "intersection".... Go Syntax doesn't support that but you can use https://pkg.go.dev/github.com/golang-collections/collections/set )


Tom

Reply:
Suggestion 1: Implemented, I have extracted the code into `DeleteUnmanagedDomains` and `DeleteUnmanagedZones`.

Regarding your problem with CSCGLOBAL and your suggestion 2:
If we have the domain `test.com` deployed on dns provider 1 and dns provider 2, but only configured for dns provider 1, it would stay deployed on dns provider 2 if we only compare to `cfg.Domains`

Therefore we need to check whether a domain is actually not managed for provider (instead of your suggestion of checking whether it is managed at all).

The code that is being used for this is:
```
func IsZoneManagedByProvider(cfg *models.DNSConfig, zone string, dnsProviderName string) bool {
	domainCfg := GetDomainCfg(cfg, zone)

	if domainCfg == nil {
		return false
	}
	for managedProviderName, _ := range domainCfg.DNSProviderNames {
		if managedProviderName == dnsProviderName {
			return true
		}
	}
	return false
}
```

For me that is working fine. Can you check on your side why this might not be working for CSCGLOBAL?

Reply:
I have just added a test `TestIsDomainOrZoneManaged`. 

To reproduce your CSCGLOBAL issue, you could try breaking it by modifying `test_data/dnsconfig.js`